Reader: Crash on search from self-hosted site

### Expected behavior
Clicking search on Reader tab should open search screen when logged in using a self-hosted site.

### Actual behavior
Clicking search on Reader tab crashes the app when logged in using a self-hosted site.

![search](https://user-images.githubusercontent.com/1405144/82558022-76413b00-9b8a-11ea-9ef5-ae050b928edd.gif)

### Steps to reproduce the behavior

1. Launch app
2. Click Log in by entering your site
3. Enter self-hosted Site address -> username/pwd
4. Click Reader tab
5. Click Search icon

Crash logs
```
java.lang.RuntimeException: Unable to start activity ComponentInfo{org.wordpress.android/org.wordpress.android.ui.reader.ReaderSearchActivity}: java.lang.NullPointerException: Attempt to invoke interface method 'boolean android.view.MenuItem.expandActionView()' on a null object reference
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2778)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'boolean android.view.MenuItem.expandActionView()' on a null object reference
        at org.wordpress.android.ui.reader.ReaderPostListFragment.onActivityCreated(ReaderPostListFragment.java:446)
        at androidx.fragment.app.Fragment.performActivityCreated(Fragment.java:2619)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:904)
        at androidx.fragment.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManagerImpl.java:1238)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:1303)
        at androidx.fragment.app.BackStackRecord.executeOps(BackStackRecord.java:439)
        at androidx.fragment.app.FragmentManagerImpl.executeOps(FragmentManagerImpl.java:2079)
        at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManagerImpl.java:1869)
        at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManagerImpl.java:1824)
        at androidx.fragment.app.FragmentManagerImpl.execPendingActions(FragmentManagerImpl.java:1727)
        at androidx.fragment.app.FragmentManagerImpl.dispatchStateChange(FragmentManagerImpl.java:2663)
        at androidx.fragment.app.FragmentManagerImpl.dispatchActivityCreated(FragmentManagerImpl.java:2613)
        at androidx.fragment.app.FragmentController.dispatchActivityCreated(FragmentController.java:246)
        at androidx.fragment.app.FragmentActivity.onStart(FragmentActivity.java:542)
        at androidx.appcompat.app.AppCompatActivity.onStart(AppCompatActivity.java:201)
        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1334)
        at android.app.Activity.performStart(Activity.java:7029)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2741)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6494) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
```

**UPDATE**: It seems we shouldn't be showing search icon in this case.

| Before | After |
| --- | --- |
|![device-2020-05-25-114404](https://user-images.githubusercontent.com/1405144/82784653-9f114b00-9e7e-11ea-8f29-af07d180f4b0.png)| ![device-2020-05-25-114453](https://user-images.githubusercontent.com/1405144/82784738-c7994500-9e7e-11ea-8729-bf9cf676aef3.png)|


##### Tested on [Pixel 3XL], Android [API 27]

--- a/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderFragment.kt
+++ b/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderFragment.kt
@@ -77,7 +77,10 @@ class ReaderFragment : Fragment(R.layout.reader_fragment_layout) {
 
     override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
         inflater.inflate(R.menu.reader_home, menu)
-        searchMenuItem = menu.findItem(R.id.menu_search)
+        menu.findItem(R.id.menu_search).apply {
+            searchMenuItem = this
+            this.isVisible = viewModel.uiState.value?.searchIconVisible ?: false
+        }
     }
 
     override fun onOptionsItemSelected(item: MenuItem): Boolean {

--- a/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderFragment.kt
+++ b/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderFragment.kt
@@ -35,6 +35,8 @@ class ReaderFragment : Fragment(R.layout.reader_fragment_layout) {
     private lateinit var viewModel: ReaderViewModel
     private lateinit var newsCardViewModel: NewsCardViewModel
 
+    private var searchMenuItem: MenuItem? = null
+
     private val viewPagerCallback = object : ViewPager2.OnPageChangeCallback() {
         override fun onPageSelected(position: Int) {
             super.onPageSelected(position)
@@ -58,6 +60,11 @@ class ReaderFragment : Fragment(R.layout.reader_fragment_layout) {
         initViewModel()
     }
 
+    override fun onDestroyView() {
+        super.onDestroyView()
+        searchMenuItem = null
+    }
+
     override fun onResume() {
         super.onResume()
         viewModel.onScreenInForeground()
@@ -70,6 +77,7 @@ class ReaderFragment : Fragment(R.layout.reader_fragment_layout) {
 
     override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
         inflater.inflate(R.menu.reader_home, menu)
+        searchMenuItem = menu.findItem(R.id.menu_search)
     }
 
     override fun onOptionsItemSelected(item: MenuItem): Boolean {
@@ -101,6 +109,7 @@ class ReaderFragment : Fragment(R.layout.reader_fragment_layout) {
         viewModel.uiState.observe(viewLifecycleOwner, Observer { uiState ->
             uiState?.let {
                 updateTabs(uiState)
+                searchMenuItem?.isVisible = uiState.searchIconVisible
             }
         })
 

--- a/WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderViewModel.kt
+++ b/WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderViewModel.kt
@@ -8,6 +8,8 @@ import kotlinx.coroutines.withContext
 import org.greenrobot.eventbus.EventBus
 import org.greenrobot.eventbus.Subscribe
 import org.greenrobot.eventbus.ThreadMode.MAIN
+import org.wordpress.android.BuildConfig
+import org.wordpress.android.fluxc.store.AccountStore
 import org.wordpress.android.models.ReaderTag
 import org.wordpress.android.models.ReaderTagList
 import org.wordpress.android.modules.BG_THREAD
@@ -34,7 +36,8 @@ class ReaderViewModel @Inject constructor(
     private val appPrefsWrapper: AppPrefsWrapper,
     private val dateProvider: DateProvider,
     private val loadReaderTabsUseCase: LoadReaderTabsUseCase,
-    private val readerTracker: ReaderTracker
+    private val readerTracker: ReaderTracker,
+    private val accountStore: AccountStore
 ) : ScopedViewModel(mainDispatcher) {
     private var initialized: Boolean = false
 
@@ -66,7 +69,8 @@ class ReaderViewModel @Inject constructor(
             if (tagList.isNotEmpty()) {
                 _uiState.value = ReaderUiState(
                         tagList.map { it.label },
-                        tagList
+                        tagList,
+                        searchIconVisible = isSearchSupported()
                 )
                 if (!initialized) {
                     initialized = true
@@ -97,7 +101,11 @@ class ReaderViewModel @Inject constructor(
         appPrefsWrapper.setReaderTag(selectedTag)
     }
 
-    data class ReaderUiState(val tabTitles: List<String>, val readerTagList: ReaderTagList)
+    data class ReaderUiState(
+        val tabTitles: List<String>,
+        val readerTagList: ReaderTagList,
+        val searchIconVisible: Boolean
+    )
 
     override fun onCleared() {
         super.onCleared()
@@ -118,7 +126,11 @@ class ReaderViewModel @Inject constructor(
     }
 
     fun onSearchActionClicked() {
-        _showSearch.value = Event(Unit)
+        if (isSearchSupported()) {
+            _showSearch.value = Event(Unit)
+        } else if (BuildConfig.DEBUG) {
+            throw IllegalStateException("Search should be hidden when isSearchSupported returns false.")
+        }
     }
 
     private fun ReaderTag.isDefaultSelectedTab(): Boolean = this.isDiscover
@@ -135,4 +147,6 @@ class ReaderViewModel @Inject constructor(
     fun onScreenInBackground() {
         readerTracker.stop(MAIN_READER)
     }
+
+    private fun isSearchSupported() = accountStore.hasAccessToken()
 }

--- a/WordPress/src/main/res/menu/reader_home.xml
+++ b/WordPress/src/main/res/menu/reader_home.xml
@@ -6,5 +6,6 @@
         android:id="@+id/menu_search"
         android:icon="@drawable/ic_search_white_24dp"
         android:title="@string/search"
+        android:visible="false"
         app:showAsAction="always" />
 </menu>

--- a/WordPress/src/test/java/org/wordpress/android/ui/reader/viewmodels/ReaderViewModelTest.kt
+++ b/WordPress/src/test/java/org/wordpress/android/ui/reader/viewmodels/ReaderViewModelTest.kt
@@ -15,6 +15,7 @@ import org.mockito.ArgumentMatchers.any
 import org.mockito.Mock
 import org.mockito.junit.MockitoJUnitRunner
 import org.wordpress.android.TEST_DISPATCHER
+import org.wordpress.android.fluxc.store.AccountStore
 import org.wordpress.android.models.ReaderTag
 import org.wordpress.android.models.ReaderTag.DISCOVER_PATH
 import org.wordpress.android.models.ReaderTag.FOLLOWING_PATH
@@ -45,6 +46,7 @@ class ReaderViewModelTest {
     @Mock lateinit var dateProvider: DateProvider
     @Mock lateinit var loadReaderTabsUseCase: LoadReaderTabsUseCase
     @Mock lateinit var readerTracker: ReaderTracker
+    @Mock lateinit var accountStore: AccountStore
 
     private val emptyReaderTagList = ReaderTagList()
     private val nonEmptyReaderTagList = ReaderTagList().apply {
@@ -62,7 +64,8 @@ class ReaderViewModelTest {
                 appPrefsWrapper,
                 dateProvider,
                 loadReaderTabsUseCase,
-                readerTracker
+                readerTracker,
+                accountStore
         )
 
         whenever(dateProvider.getCurrentDate()).thenReturn(Date(DUMMY_CURRENT_TIME))
@@ -214,6 +217,7 @@ class ReaderViewModelTest {
     @Test
     fun `OnSearchActionClicked emits showSearch event`() {
         // Arrange
+        whenever(accountStore.hasAccessToken()).thenReturn(true)
         var event: Event<Unit>? = null
         viewModel.showSearch.observeForever {
             event = it
@@ -223,6 +227,36 @@ class ReaderViewModelTest {
 
         // Assert
         assertThat(event).isNotNull
+    }
+
+    @Test
+    fun `Search is disabled for self-hosted login`() = testWithNonEmptyTags {
+        // Arrange
+        whenever(accountStore.hasAccessToken()).thenReturn(false)
+        var state: ReaderUiState? = null
+        viewModel.uiState.observeForever {
+            state = it
+        }
+        // Act
+        viewModel.start()
+
+        // Assert
+        assertThat(state!!.searchIconVisible).isFalse()
+    }
+
+    @Test
+    fun `Search is enabled for dot com login`() = testWithNonEmptyTags {
+        // Arrange
+        whenever(accountStore.hasAccessToken()).thenReturn(true)
+        var state: ReaderUiState? = null
+        viewModel.uiState.observeForever {
+            state = it
+        }
+        // Act
+        viewModel.start()
+
+        // Assert
+        assertThat(state!!.searchIconVisible).isTrue()
     }
 
     private fun <T> testWithEmptyTags(block: suspend CoroutineScope.() -> T) {

